modnet/fit_genetic.py

```diff
@@ -0,0 +1,246 @@
+        self.num_epochs = num_epochs
+        self.prob_mut = prob_mut
+        self.data = data
+
+
+    def train_val_split(
+        self,
+        data: MODData
+        )->None:
+
+        """Splits arrays or matrices into random train and validation subsets.
+        
+        Parameter:
+            data: 'MODData' data which need to be splitted.
+        """
+
+        split_border = int(0.9*data.df_targets.shape[0])
+        final_border = int(data.df_targets.shape[0])
+        self.X_train, self.X_val = data.split((range(split_border),range(split_border,final_border)))
+        self.y_train = self.X_train.df_targets
+        self.y_val = self.X_val.df_targets
+
+        return self.X_train, self.X_val, self.y_train, self.y_val
+
+
+    def initialization_population(
+        self,
+        size_pop: int
+        )->None:
+
+        """Inintializes the initial population (Generation 0).
+       
+        Paramter:
+            size_pop: Size of the population.
+        """
+
+        self.pop =  [[]]*size_pop
+        activation = ['relu', 'elu']
+        loss = ['mae', 'mse']
+        xscale = ['minmax', 'normal']
+        lr = [0.03, 0.01, 0.003, 0.001, 0.0003, 0.0001]
+        initial_batch_size = [8, 16, 32, 64, 128]
+        fraction = [1, 0.75, 0.5, 0.25]
+        self.pop = [[10*randint(1,int(len(self.X_train.get_optimal_descriptors())/10)), 32*randint(1,10), random.choice(fraction), random.choice(fraction), random.choice(fraction), random.choice(activation), random.choice(loss), random.choice(xscale), random.choice(lr), random.choice(initial_batch_size)] for i in range(0, size_pop)]
+        return self.pop
+
+
+    def crossover(
+        self,
+        mother: List,
+        father: List
+        )->None:
+
+        """Does the crossover of two parents and returns a 'child' which have the combined genetic information of both parents.
+
+        Parameters:
+            mother: List containing the gentic information of the first parent.
+            father: List containing the gentic information of the second parent.
+        """
+
+        child = [mother[0], father[1], mother[2], father[3], mother[4], father[5], mother[6], father[7], mother[8], father[9]]   
+        return child
+
+
+    def mutation(
+        self,
+        child: List,
+        prob_mut: float = 0.8
+        )->None:
+
+        """Performs mutation in the genetic information in order to maintain diversity in the population. 
+        
+        Paramters:
+            child: List containing the genetic information of the 'child'.
+            prob_mut: Probability the mutation occurs.
+        """
+
+        for c in range(0, len(child)):
+            if np.random.rand() > prob_mut:
+                if child[c][0] < int(0.9*len(self.X_train.get_optimal_descriptors())):
+                    child[c][0] = int(child[c][0] + 10*randint(1, 6))
+                else:
+                    child[c][0] = int(child[c][0] - 10*randint(1, 6))
```

Comment:
The random offset on `n_feat` should probably be centred on 0, but shouldn't the rest of the genome also be altered?

I would consider making functions that mutate the overall architecture (i.e. make every layer wider or thinner, make network narrow more quickly/slowly)